frontend/src/components/ModelDiff.jsx

```diff
@@ -2,88 +2,104 @@
+    head: false,
+    details: false
+  });
+  const [diffSuccess, setDiffSuccess] = useState(null);
 
   const handleSelectChange = (e) => {
-    const option = e.target.value;
-    setSelectedOption(option);
-    const selectedValue = JSON.parse(e.target.value);
-    setSelectedDetails(selectedValue);
-    setIsExpanded(false);
+    const option = JSON.parse(e.target.value);
+    setSelectedOption(e.target.value);
+    setSelectedDetails(option);
+    setIsExpanded((prev) => ({ ...prev, details: false }));
   };
 
   const handleGenerateDiff = async () => {
-    if (!commitDetails[0].hash || !selectedDetails.hash) {
+    const headCommit = commitDetails[0]?.hash;
+    const selectedCommit = selectedDetails?.hash;
+
+    if (!headCommit || !selectedCommit) {
       alert('Please select a version.');
       return;
     }
-    setCompleteLoading(false);
-    setIsLoading(true);
+
+    setLoadingState('loading');
+    setError('');
+
     try {
-      const url = API_BASE_URL + '/compare';
-      const response = await postData(url, {
-        head: commitDetails[0].hash,
-        prev: selectedDetails.hash
+      const response = await postData(`${API_BASE_URL}/compare`, {
+        head: headCommit,
+        prev: selectedCommit
       });
       const data = await response.json();
+
       if (data.error) {
         throw new Error(data.error);
       }
+
+      setDiffSuccess(true);
+      setLoadingState('complete');
     } catch (error) {
-      console.error('Error:', error);
+      setDiffSuccess(false);
+      setError(error.message);
+      setLoadingState('error');
     } finally {
-      setIsLoading(false);
-      setCompleteLoading(true);
       onRefetch();
     }
   };
 
-  const postData = async (url = '', data = {}) => {
+  const postData = async (url, data) => {
     try {
       const response = await fetch(url, {
         method: 'POST',
-        headers: {
-          'Content-Type': 'application/json'
-        },
+        headers: { 'Content-Type': 'application/json' },
         body: JSON.stringify(data)
       });
-      if (!response.ok) {
-        throw new Error('Network response was not ok');
-      }
+
+      if (!response.ok) throw new Error('Network response was not ok');
       return response;
     } catch (error) {
       console.error('Error in postData:', error);
       throw error;
     }
   };
 
-  async function openModelCompareDialog() {
+  const openModelCompareDialog = async () => {
     try {
-      const response = await fetch(API_BASE_URL + '/commits');
-      if (!response.ok) {
-        throw new Error(
-          'Failed to fetch commits info: ' + response.statusText
-        );
-      }
+      setError('');
+      setSelectedOption('');
+      setSelectedDetails(null);
+      setDiffSuccess(null);
+      setLoadingState('idle');
+
+      const response = await fetch(`${API_BASE_URL}/commits`);
+
+      if (!response.ok) throw new Error('Failed to fetch commits info.');
```

Comment:
This doesn't seem to properly handle an error response from the server. When I click the button, but the server responds with a 500 error, nothing visibly happens.